nixos/modules/system/boot/grow-partition.nix

```diff
@@ -16,29 +16,22 @@ with lib;
   };
 
   config = mkIf config.boot.growPartition {
-
-    assertions = [{
-      assertion = !config.boot.initrd.systemd.enable;
-      message = "systemd stage 1 does not support 'boot.growPartition' yet.";
-    }];
-
-    boot.initrd.extraUtilsCommands = ''
-      copy_bin_and_libs ${pkgs.gawk}/bin/gawk
-      copy_bin_and_libs ${pkgs.gnused}/bin/sed
-      copy_bin_and_libs ${pkgs.util-linux}/sbin/sfdisk
-      copy_bin_and_libs ${pkgs.util-linux}/sbin/lsblk
-
-      substitute "${pkgs.cloud-utils.guest}/bin/.growpart-wrapped" "$out/bin/growpart" \
-        --replace "${pkgs.bash}/bin/sh" "/bin/sh" \
-        --replace "awk" "gawk" \
-        --replace "sed" "gnused"
-
-      ln -s sed $out/bin/gnused
-    '';
-
-    boot.initrd.postDeviceCommands = ''
-      rootDevice="${config.fileSystems."/".device}"
-      if waitDevice "$rootDevice"; then
+    systemd.services.growpart = {
+      wantedBy = [ "-.mount" ];
+      after = [ "-.mount" "systemd-repart.service" ];
+      before = [ "systemd-growfs-root.service" "systemd-repart.service" ]; # such that not two things write to the partition table at the same time
+      conflicts = [ "shutdown.target" ];
+      unitConfig.DefaultDependencies = false;
+      # growpart returns 1 if the partition is already grown
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+        TimeoutSec = "infinity";
+        SuccessExitStatus = "0 1";
+      };
+
+      script = ''
+        rootDevice="${config.fileSystems."/".device}"
```

Comment:
I wonder if we should do like systemd does and use something like `stat --format="%t %T" /` to just get the major/minor device numbers of whatever is mounted on root, rather than pulling the device path from the nix configuration? Would have the added benefit of working with stuff like `device = "LABEL=foo";`.

Not really a big deal; just a thought I had.

Reply:
I'm gonna keep the behavior as is I think. We could use `findmnt` or something but I don't really care myself as I'm using repart. (But that only works for GPT and not MBBR and most our cloud images are MBR)